### PR TITLE
Bring back loss information for multiple outputs

### LINKED ISSUE #18993: loss tracker for multiple model outputs
Currently, only the total loss has been tracked and displayed.
Can we have more detailed information for each loss?

I have a workaround to recompute the loss using `metrics`, but it seems very inefficient.

### LINKED ISSUE #19219: [Feature Request] Re-enable multi-output logs
If a model is configured for multi-output, in keras 2, we can get the log message for all corresponding outputs. But in keras 3, it is removed and only shows the summarized log message. Unlike mult-class classification, returning log message for mult-output sounds more natural. 

```python
priority_pred = layers.Dense(1, name="priority")(x)
department_pred = layers.Dense(num_departments, name="department")(x)

model = keras.Model(
    inputs=[title_input, body_input, tags_input],
    outputs={"priority": priority_pred, "department": department_pred},
)

model.compile(
    optimizer=keras.optimizers.RMSprop(1e-3),
    loss={
        "priority": keras.losses.BinaryCrossentropy(from_logits=True),
        "department": keras.losses.CategoricalCrossentropy(from_logits=True),
    },
    loss_weights={"priority": 1.0, "department": 0.2},
)

model.fit
```

in keras 2

```python
Epoch 1/2
 40/40 ━ 1s 12ms/step - loss: ?, priority_loss: ?, department_loss: ?
```

in keras 3

```python
Epoch 1/2
 40/40 ━ 1s 12ms/step - loss: ?
```

Posted as a bug https://github.com/keras-team/keras/issues/19159, but it turned out that it was intentionally removed.

### PR DESCRIPTION
**Pull request by james77777778**
Closes #18993 
Closes #19219
Related to #18467

This is an important feature, especially for a multi-modal model, that is currently missing in Keras 3 compared to Keras 2.
The implementation is quite simple by adding the `Mean` metric in `CompileLoss`.

All TODOs in `model_tests.py` have been resolved.

A simple demo:
```python
import numpy as np

from keras import backend
from keras import layers
from keras import models

np.random.seed(42)
x = np.random.rand(8, 3)
y1 = np.random.rand(8, 1)
y2 = np.random.randint(0, 2, (8, 1))

inputs = layers.Input(shape=(3,), name="inputs")
output_a = layers.Dense(1, name="a")(inputs)
output_b = layers.Dense(1, name="b")(inputs)
model = models.Model(inputs, {"a": output_a, "b": output_b})
init_weights = model.get_weights()
model.compile(
    optimizer="sgd",
    loss={"a": "mean_squared_error", "b": "binary_crossentropy"},
)
if backend.backend() != "numpy":
    model.fit(x, (y1, y2), batch_size=2, epochs=1, shuffle=False, verbose=1)
model.evaluate(x, (y1, y2), batch_size=2, verbose=1)

```
Got this:
```bash
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - a_loss: 0.6758 - b_loss: 6.0316 - loss: 6.7074
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - a_loss: 0.5584 - b_loss: 6.0045 - loss: 6.5630
```

EDITED:
I have introduced a new scope (`SymbolicScope`) to control the behavior in `compute_loss`. The reason is that `self.losses` will break when using `backend.compute_output_spec`. Some edge cases in CI also require this.

### CODE REVIEW

**Comment by james77777778** [comment]
Still some bugs need to be resolved.
Will update this PR tomorrow.

**Comment by james77777778** [comment]
This PR should be ready now.

File: keras/src/trainers/trainer.py [hunk 1]
```diff
@@ -325,8 +328,12 @@ def metrics(self):
             loss = self._compile_loss(y, y_pred, sample_weight)
             if loss is not None:
                 losses.append(loss)
-        for loss in self.losses:
-            losses.append(ops.sum(ops.cast(loss, dtype=backend.floatx())))
+
+        # If in symbolic scope, skip `self.losses` to ensure we don't access
+        # any variables. Otherwise, it might break.
+        if not in_symbolic_scope():
```
Comment (fchollet) [inline keras/src/trainers/trainer.py:334]: This feels quite brittle -- `compute_loss` is a method that is intended to be overridden by users in various cases, and those users should not need to be aware of `SymbolicScope`. Why is it necessary?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

> This is an important feature, especially for a multi-modal model

What's wrong with just letting users specify what they want to monitor via the `metrics` argument?

I think that if we're going to do this, we should find a solution that only adds a minimal amount of complexity, e.g. if the loss is non-unary, add the loss configuration to `CompileMetrics`.

**Comment by james77777778** [comment]
> What's wrong with just letting users specify what they want to monitor via the `metrics` argument?

It is fine if the loss computation is not intensive; the redundant loss computation in `CompileMetrics` is negligible.

```python
import numpy as np

from keras import backend
from keras import layers
from keras import models

np.random.seed(42)
x = np.random.rand(8, 3)
y1 = np.random.rand(8, 1)
y2 = np.random.randint(0, 2, (8, 1))

inputs = layers.Input(shape=(3,), name="inputs")
output_a = layers.Dense(1, name="a")(inputs)
output_b = layers.Dense(1, name="b")(inputs)
model = models.Model(inputs, {"a": output_a, "b": output_b})
model.compile(
    optimizer="sgd",
    loss={"a": "mean_squared_error", "b": "binary_crossentropy"},
    metrics={"a": "mean_squared_error", "b": "binary_crossentropy"},
)
if backend.backend() != "numpy":
    model.fit(x, (y1, y2), batch_size=2, epochs=1, shuffle=False, verbose=1)
logs = model.evaluate(x, (y1, y2), batch_size=2, verbose=1, return_dict=True)

```

However, things become noticeably complicated if we override `compute_loss` for an intensive loss computation like this one in KerasCV:
https://github.com/keras-team/keras-cv/blob/master/keras_cv/src/models/object_detection/yolo_v8/yolo_v8_detector.py#L534-L584
There is a `YOLOV8LabelEncoder`, which is computation-intensive, in `compute_loss`.

Additionally, it is inconvenient to require users to implement almost the same logic in `compute_metrics`.

> I think that if we're going to do this, we should find a solution that only adds a minimal amount of complexity, e.g. if the loss is non-unary, add the loss configuration to `CompileMetrics`.

I don't think adding the loss config to `CompileMetrics` resolve the issue of redundant loss computation, does it?

File: keras/src/trainers/trainer.py (hunk 1 above)
Comment (james77777778) [reply keras/src/trainers/trainer.py]: I have moved the logic to `Layer._get_regularization_losses`.
This should be more robust now.

**Comment by james77777778** [comment]
@fchollet 
I just spotted a missing `training` bug in `TorchWrapper` and fixed it. The corresponding tests have been included.

Please let me know if this PR is a good fit or not for Keras3.
If we don't want this multiple logs feature, I can separate the `TorchWrapper` fix into another PR.

**Review by fchollet — APPROVED** [review]
With the latest update, I think this change can be merged. Thank you for the contribution!

**Comment by ghsanti** [comment]
this is just a random comment for the most part but shouldn't 

`utils.set_random_seed()` be used instead of numpy's random seed since the `Dense` layers' kernel initializer will not allow for reproducibility otherwise unless explicitly added.

**Comment by james77777778** [comment]
Hey @ghsanti 
I'm unsure what you mean.
If you are talking about the code for the simple demo, it's just a demo to showcase this PR.

**Comment by ghsanti** [comment]
> Hey @ghsanti I'm unsure what you mean. If you are talking about the code for the simple demo, it's just a demo to showcase this PR.

i see it was for showing the losses, ty for the reply!

**Comment by early-stopper** [comment]
This covers the case that a single loss is assigned to each output. How would I do this in case of a model with more losses than outputs? For instance, a variational autoencoder that has a single output (the decoded image), but multiple losses (reconstruction loss + KL loss)? How can I make sure that bothlosses are logged separately?
